### PR TITLE
Add: Added the instructions for debugging from rust-forge

### PR DESCRIPTION
**Pull request by cg-cnu**
* Added the instructions for debugging in `debugging.md` 
* Added a link to `SUMMARY.md` at the end of the file

Let me know if you got any comments. Thanks 🙂

### CODE REVIEW

**Comment by cg-cnu** [comment]
Working on the issue #11

File: src/compiler-debugging.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+   ```
+stack backtrace:
+   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
+   1: std::sys_common::backtrace::_print
+   2: std::panicking::default_hook::{{closure}}
+   3: std::panicking::default_hook
+   4: std::panicking::rust_panic_with_hook
+   5: std::panicking::begin_panic
+   6: rustc_typeck::check::cast::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, \
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: Typo?

File: src/compiler-debugging.md [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+When you have an ICE (panic in the compiler), you can set
+`RUST_BACKTRACE=1` to get the stack trace of the `panic!` like in
+normal Rust programs.  IIRC backtraces **don't work** on Mac and on MinGW,
+sorry. If you have trouble or the backtraces are full of `unknown`,
+you might want to find some way to use Linux or MSVC on Windows.
+
+In the default configuration, you don't have line numbers enabled, so the 
+backtrace looks like this:
+
+   ```
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: Should be ` ```text `

File: src/SUMMARY.md [hunk 3]
```diff
@@ -52,6 +52,7 @@
     - [miri const evaluator](./miri.md)
 - [Parameter Environments](./param_env.md)
 - [Generating LLVM IR](./trans.md)
+- [Debugging the Compiler](./compiler-debugging.md)
```
Comment (mark-i-m) [inline src/SUMMARY.md]: I think this should be moved after the testing chapter (ch 4)

File: src/compiler-debugging.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+[debugging]: #debugging
+
+Here are a few tips to debug the compiler:
+
+## Getting a backtrace
+[getting-a-backtrace]: #getting-a-backtrace
+
+When you have an ICE (panic in the compiler), you can set
+`RUST_BACKTRACE=1` to get the stack trace of the `panic!` like in
+normal Rust programs.  IIRC backtraces **don't work** on Mac and on MinGW,
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: @nikomatsakis Is this still true?

File: src/compiler-debugging.md (hunk 1 above)
Comment (mark-i-m) [reply src/compiler-debugging.md]: If this line is too long, then we could just remove it altogether...

File: src/compiler-debugging.md [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+   ```
+
+If you want line numbers for the stack trace, you can enable 
+`debuginfo-lines=true` or `debuginfo=true` in your config.toml and rebuild the 
+compiler. Then the backtrace will look like this:
+
+```
+stack backtrace:
+   (~~~~ LINES REMOVED BY ME FOR BREVITY ~~~~)
+   6: rustc_typeck::check::cast::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, \
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: same as above

File: src/compiler-debugging.md [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+  35: rustc::ty::context::TyCtxt::create_and_enter
+  36: rustc_driver::driver::compile_input
+  37: rustc_driver::run_compiler
+   ```
+
+If you want line numbers for the stack trace, you can enable 
+`debuginfo-lines=true` or `debuginfo=true` in your config.toml and rebuild the 
+compiler. Then the backtrace will look like this:
+
+```
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: should be ` ```text `

File: src/compiler-debugging.md [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+If you want to get a backtrace to the point where the compiler emits
+an error message, you can pass the `-Z treat-err-as-bug`, which
+will make the compiler panic on the first error it sees.
+
+This can also help when debugging `delay_span_bug` calls - it will make
+the first `delay_span_bug` call panic, which will give you a useful backtrace.
+
+For example:
+```
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: should be ` ```text`

File: src/compiler-debugging.md [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+             at /home/user/rust/src/librustc_driver/lib.rs:253
+$ # Cool, now I have a backtrace for the error
+```
+
+## Getting logging output
+[getting-logging-output]: #getting-logging-output
+
+The compiler has a lot of `debug!` calls, which print out logging information
+at many points. These are very useful to at least narrow down the location of
+a bug if not to find it entirely, or just to orient yourself to why a compiler
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: s/to orient yourself to why a compiler/to orient yourself as to why the compiler/

File: src/compiler-debugging.md [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+powered by [env-logger], and you can look at the docs linked there to see
+the full `RUST_LOG` syntax. All `debug!` output will then appear in
+standard error.
+
+Note that unless you use a very strict filter, the logger will emit a *lot*
+of output - so it's typically a good idea to pipe standard error to a file
+and look at the log output with a text editor.
+
+So to put it together.
+```
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: should be ` ```text `

File: src/compiler-debugging.md [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+config.toml.
+
+I also think that in some cases just setting it will not trigger a rebuild,
+so if you changed it and you already have a compiler built, you might
+want to call `x.py clean` to force one.
+
+### Logging etiquette
+
+Because calls to `debug!` are removed by default, in most cases, don't worry
+about adding "unnecessary" calls to `debug!` and leaving them in in code
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: typo: "in in"

File: src/compiler-debugging.md [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+[formatting-graphviz-output]: #formatting-graphviz-output
+
+Some compiler options for debugging specific features yield graphviz graphs - 
+e.g. the `#[rustc_mir(borrowck_graphviz_postflow="suffix.dot")]` attribute
+dumps various borrow-checker dataflow graphs.
+
+These all produce `.dot` files. To view these files, install graphviz (e.g.
+`apt-get install graphviz`) and then run the following commands:
+
+```
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: I think ` ```bash `?

File: src/compiler-debugging.md [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+
+The rustc build process builds the LLVM tools into 
+`./build/<host-triple>/llvm/bin`. They can be called directly.
+
+The default rustc compilation pipeline has multiple codegen units, which is hard
+to replicate manually and means that LLVM is called multiple times in parallel.
+If you can get away with it (i.e. if it doesn't make your bug disappear),
+passing `-C codegen-units=1` to rustc will make debugging easier.
+
+If you want to play with the optimization pipeline, you can use the `opt` from
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: Perhaps rephrase this sentence as "If you want to play with the optimization pipeline, you can use the `opt` tool from `./build/<host-triple>/llvm/bin/` with the the LLVM IR emitted by rustc."

File: src/compiler-debugging.md [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+there on the IR rustc emits with `--emit=llvm-ir`. Note
+that rustc emits different IR depending on whether `-O` is enabled, even without
+LLVM's optimizations, so if you want to play with the IR rustc emits, 
+you should:
+```
```
Comment (mark-i-m) [inline src/compiler-debugging.md]: ` ```bash `

File: src/compiler-debugging.md (hunk 1 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: I had to reduce the line length, so kept it in a separate line. Now removed it.

File: src/compiler-debugging.md (hunk 2 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: made it text

File: src/SUMMARY.md (hunk 3 above)
Comment (cg-cnu) [reply src/SUMMARY.md]: Done!

File: src/compiler-debugging.md (hunk 5 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: removed the line and made it ```text

File: src/compiler-debugging.md (hunk 6 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: done!

File: src/compiler-debugging.md (hunk 7 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: Its shell commands and comments, so made it bash! Hope its fine!

File: src/compiler-debugging.md (hunk 8 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: changed it!

File: src/compiler-debugging.md (hunk 9 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: again, couple of shell commands and comments, so made it ```bash

File: src/compiler-debugging.md (hunk 10 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: Fixed

File: src/compiler-debugging.md (hunk 11 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: yep! Made it bash!

File: src/compiler-debugging.md (hunk 12 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: Changed the first sentence and ended up with this paragraph! 

>If you want to play with the optimization pipeline, you can use the opt tool 
from `./build/<host-triple>/llvm/bin/` with the the LLVM IR emitted by rustc. 
Note that rustc emits different IR depending on whether `-O` is enabled, even 
without LLVM's optimizations, so if you want to play with the IR rustc emits, 
you should:

let me know if it's fine!

File: src/compiler-debugging.md (hunk 13 above)
Comment (cg-cnu) [reply src/compiler-debugging.md]: done!

**Comment by cg-cnu** [comment]
Hey @mark-i-m Fixed all the suggested changes and travis build is passing now! Waiting on one comment from @nikomatsakis Let me know if you have any other suggestions. Thanks! 🙂

**Comment by mark-i-m** [comment]
@cg-cnu Thanks :)

Let's wait a little while, and if Niko can't get around to it, we can add a note like "Note: this is copied from the forge. If anything needs updating, please open an issue or make a PR on the github repo." and merge...

**Comment by cg-cnu** [comment]
Thanks @mark-i-m ! I just did a PR on rust-forge to remove `debugging.md` https://github.com/rust-lang-nursery/rust-forge/pull/134

**Comment by mark-i-m** [comment]
Ok, so Niko is going on (well-earned) vacation. Could add a note at the top of the document (maybe bolded or something)? Then I think we can just go ahead and merge.

**Comment by cg-cnu** [comment]
Hey @mark-i-m Added note at the top of the page in bold! 🙂

**Comment by mark-i-m** [comment]
Thanks!

**Pull request merged**